chain/types/ethtypes/eth_types.go

```diff
@@ -529,7 +529,13 @@ func handleHexStringPrefix(s string) string {
 }
 
 func EthHashFromCid(c cid.Cid) (EthHash, error) {
-	return ParseEthHash(c.Hash().HexString()[8:])
+	b := c.Hash()
+	if len(b) != EthHashLength+4 {
+		return EthHash{}, fmt.Errorf("CID hash length is not 32 bytes")
+	}
```

Comment:
I'd still check to make sure it's the right type. Really, I'd:

1. Use `cid.Prefix` to compute the expected byte-prefix. Store that in a const/global var.
2. Use `c.Bytes()` to get the bytes.
3. Use `bytes.CutPrefix` to remove/check for this prefix.
4. Check that the remaining digest is 32 bytes long.

That'll avoid _all_ parsing.

Reply:
Hm. I guess we weren't even checking that before...

Reply:
E.g., see https://github.com/filecoin-project/lotus/blob/e0c6047bf20cbe9d68b595a9a4d0080c218d605b/chain/sub/incoming.go#L40-L47

Reply:
cool, I didn't know about `CutPrefix`, but also this does make it stricter than it currently is, we're OK with that? It may not make sense to do it any other way but you can currently feed it a CID with the wrong codec and hash function as long as the digest is what you want and it'll work.

Reply:
> we're OK with that?

IMO, the current lax-ness is a bug.

Reply:
The tests disagree with me... is that an issue with the tests?

Reply:
Yeah, we're using a fake CID.

Reply:
defaked

Reply:
well, better faked